kernel/scheduler/src/lib.rs

```diff
@@ -1,82 +1,158 @@
 
-/// Yields the current CPU by selecting a new `Task` to run 
+fn get_least_busy_run_queue() -> Option<&'static RwLockPreempt<RunQueue>> {
+    let mut min_rq: Option<(&'static RwLockPreempt<RunQueue>, usize)> = None;
+
+    for (_, rq) in RUN_QUEUES.iter() {
+        let rq_size = rq.read().len();
+
+        if let Some(min) = min_rq {
+            if rq_size < min.1 {
+                min_rq = Some((rq, rq_size));
+            }
+        } else {
+            min_rq = Some((rq, rq_size));
+        }
+    }
+
+    min_rq.map(|m| m.0)
+}
+
+#[cfg(runqueue_spillful)]
+/// Removes a `TaskRef` from the RunQueue(s) on the given `core`.
+/// Note: This method is only used by the state spillful runqueue
+/// implementation.
+pub fn remove_task_from_within_task(task: &TaskRef, core: u8) -> Result<(), &'static str> {
+    task.set_on_runqueue(None);
+    RUN_QUEUES
+        .get(&core)
+        .ok_or("Couldn't get runqueue for specified core")
+        .and_then(|rq| {
+            // Instead of calling `remove_task`, we directly call `remove_internal`
+            // because we want to actually remove the task from the runqueue,
+            // as calling `remove_task` would do nothing due to it skipping the actual
+            // removal when the `runqueue_spillful` cfg is enabled.
+            rq.write().remove_internal(task)
+        })
+}
+
+/// Yields the current CPU by selecting a new `Task` to run
 /// and then switching to that new `Task`.
 ///
 /// Preemption will be disabled while this function runs,
 /// but interrupts are not disabled because it is not necessary.
 ///
 /// ## Return
 /// * `true` if a new task was selected and switched to.
-/// * `false` if no new task was selected,
-///    meaning the current task will continue running.
+/// * `false` if no new task was selected, meaning the current task will
+///   continue running.
 pub fn schedule() -> bool {
     let preemption_guard = preemption::hold_preemption();
     // If preemption was not previously enabled (before we disabled it above),
     // then we shouldn't perform a task switch here.
     if !preemption_guard.preemption_was_enabled() {
-        // trace!("Note: preemption was disabled on CPU {}, skipping scheduler.", current_cpu());
         return false;
     }
 
-    let cpu_id = preemption_guard.cpu_id();
-
-    let Some(next_task) = scheduler::select_next_task(cpu_id) else {
-        return false; // keep running the same current task
+    let apic_id = preemption_guard.cpu_id();
+    let mut run_queue = match RUN_QUEUES.get(&apic_id) {
+        Some(rq) => rq.write(),
+        _ => {
+            error!(
+                "BUG: schedule(): couldn't get run queue for core {}",
+                apic_id
+            );
+            return false;
+        }
     };
 
-    let (did_switch, recovered_preemption_guard) = task::task_switch(
-        next_task,
-        cpu_id,
-        preemption_guard,
-    ); 
+    let next_task = run_queue.next();
+    drop(run_queue);
```

Comment:
i have experienced weird behavior before where `drop`ping a lock guard does not actually drop it until the end of the lexical scope. Therefore, I'd prefer to surround the guard with curly brackets (from which you can "return" `next_task`) to ensure that happens.